code/modules/mob/living/silicon/ai/life.dm

```diff
@@ -1,171 +1,181 @@
+	malfhacking = 0
+	malfhack = null
+
+/mob/living/silicon/ai/proc/life_handle_power_damage()
+	if(aiRestorePowerRoutine != 0)
+		// Lost power
+		adjustOxyLoss(1)
+	else
+		// Gain Power
+		adjustOxyLoss(-1)
+
+/mob/living/silicon/ai/proc/life_handle_powered_core()
+	var/unblindme = FALSE
+	if(client && client.eye == eyeobj) // We are viewing the world through our "eye" mob.
+		change_sight(adding = SEE_TURFS|SEE_MOBS|SEE_OBJS)
+		see_in_dark = 8
+		see_invisible = SEE_INVISIBLE_LEVEL_TWO
+
+	var/area/home = get_area(src)
+	if(home && home.powered(EQUIP))
+		home.use_power(1000, EQUIP)
+
+	if (aiRestorePowerRoutine==2)
+		to_chat(src, "Alert cancelled. Power has been restored without our assistance.")
+		aiRestorePowerRoutine = 0
+		unblindme = TRUE
+	else if (aiRestorePowerRoutine==3)
+		to_chat(src, "Alert cancelled. Power has been restored.")
+		aiRestorePowerRoutine = 0
+		unblindme = TRUE
+	else if (aiRestorePowerRoutine == -1)
+		to_chat(src, "Alert cancelled. External power source detected.")
+		aiRestorePowerRoutine = 0
+		unblindme = TRUE
+	if(unblindme)
+		clear_fullscreen("blind")
+	return TRUE
+
+/mob/living/silicon/ai/proc/power_restored()
+	to_chat(src, "Alert cancelled. Power has been restored without our assistance.")
+	aiRestorePowerRoutine = 0
+	clear_fullscreen("blind")
+
+/mob/living/silicon/ai/proc/life_handle_unpowered_core()
+	overlay_fullscreen("blind", /obj/abstract/screen/fullscreen/blind)
+	if(client)
+		change_sight(removing = SEE_TURFS|SEE_MOBS|SEE_OBJS)
+		see_in_dark = 0
+	see_invisible = SEE_INVISIBLE_LIVING
+
+	if (aiRestorePowerRoutine != 0)
+		return
+	aiRestorePowerRoutine = 1
+	to_chat(src, "You've lost power!")
+	spawn(2 SECONDS)
+		if(!aiRestorePowerRoutine)
+			return // Checking for premature changes.
+		to_chat(src, "Backup battery online. Scanners, camera, and radio interface offline. Beginning fault-detection.")
+		sleep(5 SECONDS)
+		if(!aiRestorePowerRoutine)
+			return // Checking for premature changes.
+		if(is_ai_powered())
+			power_restored()
+			return
+		to_chat(src, "Fault confirmed: missing external power. Shutting down main control system to save power.")
+		sleep(2 SECONDS)
+		if(!aiRestorePowerRoutine)
+			return // Checking for premature changes.
+		to_chat(src, "Emergency control system online. Verifying connection to power network.")
+		sleep(5 SECONDS)
+		if(!aiRestorePowerRoutine)
+			return // Checking for premature changes.
+		if (istype(get_turf(src), /turf/space))
+			to_chat(src, "Unable to verify! No power connection detected!")
+			aiRestorePowerRoutine = 2
+			return
+		to_chat(src, "Connection verified. Searching for APC in power network.")
+		sleep(5 SECONDS)
+		if(!aiRestorePowerRoutine)
+			return // Checking for premature changes.
+		var/obj/machinery/power/apc/theAPC = null
+
+		var/PRP //like ERP with the code, at least this stuff is no more 4x sametext
+		for (PRP=1, PRP<=4, PRP++)
```

Comment:
So with the way this whole loop works, would an AI that got carded mid-power outage (if that's even possible I can't remember) die of power loss?

Reply:
n-no, I'm fairly sure that's gonna be handled correctly
during the next Life() `aiRestorePowerRoutine` will be set to 0 and this loop should abort
I'll test it tomorrow if that doesn't sound right

Reply:
This works fine.